### ISSUE TITLE
Last Lap time and speed are wrong

### ISSUE THREAD

**Issue by The-BDD-Coach** [ssaring/sportstracker#266]
I went for a run on the track, and ran the last lap noticeably faster than the previous laps. After the run:
  - 'Main' tab shows 29:04 minutes and 7:17 per mile
  - HRM 'Optional' tab shows 7:18 per mile
  - 'Laps' tab shows the last lap split time of 00:31:15.0
  - 'Samples' tab shows the run ending at 00:29:05
  - 'Samples' tab shows every sample in the last lap faster than a 7:20 pace. This is consistent with what I was seeing on the watch during the last lap.
  - 'Laps' tab shows the last lap average speed as 7:50 min/mile
  - 'Diagram' tab shows the last lap at all times faster than 7:20 min/mile

I am using a Garmin Forerunner 245 and SportsTracker 8.0.1.
![Screenshot 2024-02-20 at 8 11 42 PM](https://github.com/ssaring/sportstracker/assets/118138954/29c6febf-d40f-4b45-b68f-524a24e8d3ff)
![Screenshot 2024-02-20 at 8 12 16 PM](https://github.com/ssaring/sportstracker/assets/118138954/3f38057e-430d-4723-99c8-86a79944eee8)
![Screenshot 2024-02-20 at 8 23 44 PM](https://github.com/ssaring/sportstracker/assets/118138954/64861b83-ff53-454b-856f-525bd8bcbe6d)

I looked at other runs and saw the same pattern - the last lap was consistently 2 to 4 minutes longer in the 'Laps' tab than in the 'Samples' tab, and of course the last lap speed was slower.

**Comment by ssaring** [comment]
Interesting. Could you please attach the recorded FIT file for the issues and screenshots reported above? Then it's easier to take a look on that.

**Comment by The-BDD-Coach** [comment]
Please let me know if you need anything else.

 

Leslie Brooks

770-841-8402

----------------------

 

From: Stefan Saring ***@***.***>
Reply-To: ssaring/sportstracker ***@***.***>
Date: Wednesday, February 21, 2024 at 12:15 PM
To: ssaring/sportstracker ***@***.***>
Cc: The-BDD-Coach ***@***.***>, Author ***@***.***>
Subject: Re: [ssaring/sportstracker] Last Lap time and speed are wrong (Issue #266)

 

Interesting. Could you please attach the recorded FIT file for the issues and screenshots reported above? Then it's easier to take a look on that.

**Comment by ssaring** [comment]
I'm not able to find any FIT file in this issue report. Perhaps the upload has failed, or I simply missed it.
In case it's not possible to upload, please send it me via mail to the contact address shown in the About dialog of SportsTracker.

**Comment by The-BDD-Coach** [comment]
I’m sorry; I replied to your previous email and attached the FIT file. I will send it again.

 

Leslie Brooks

770-841-8402

----------------------

 

From: Stefan Saring ***@***.***>
Reply-To: ssaring/sportstracker ***@***.***>
Date: Thursday, February 22, 2024 at 2:39 PM
To: ssaring/sportstracker ***@***.***>
Cc: The-BDD-Coach ***@***.***>, Author ***@***.***>
Subject: Re: [ssaring/sportstracker] Last Lap time and speed are wrong (Issue #266)

 

I'm not able to find any FIT file in this issue report. Perhaps the upload has failed, or I simply missed it.
In case it's not possible to upload, please send it me via mail to the contact address shown in the About dialog of SportsTracker.

**Comment by ssaring** [comment]
Just took a look on the FIT file you sent, that's interesting.

To answer your first question:

> 'Main' tab shows 29:04 minutes and 7:17 per mile
> HRM 'Optional' tab shows 7:18 per mile

This will be caused by rounding and conversation. The main tab in the SportsTracker Exercise dialog stores the AVG speed / pace value once you import the FIT file. In your case you also selected "Calculate automatically" for the AVG speed. Therefore it calculates it from the imported distance and duration, which might be rounded in SportsTracker with slight differences. In addition SportsTracker uses internally the metric unit system, if the user wants to see english units (miles), the data will be converted on the fly.

In summary: there might be a tiny difference of the speed / pace value stored and displayed in SportsTracker compared to the data shown in the ExerciseViewer dialog (which is displaying only the data stored in the FIT file).
I would not like to see this as an error. I also have often the same tiny difference even on my Garmin watch. E.g. the AVG pace is slightly different when stopping a run compared to the AVG speed in the activity protocol afterwards.

The problem with the duration of the last lap seems to be more of an issue, I'll take a look on that later on.

**Comment by The-BDD-Coach** [comment]
I figured that 7:17 versus 7:18 was probably a precision and rounding issue, but included that for completeness in the bug report since the overall problem is duration and speed.

**Comment by ssaring** [comment]
Found the actual problem in lap time calculation and so also in speed derivation.
This issue occurs for most Garmin devices when parsing the recorded FIT data, I've detected the same problem in my running / cycling files recorded with a FR 955.

This bug is solved now, the fix has been pushed to the master branch.
Thanks for reporting.
